cli/src/cast_opts.rs

```diff
@@ -252,19 +253,44 @@ pub struct EthereumOpts {
     pub wallet: Wallet,
 }
 
-// TODO: Improve these so that we return a middleware trait object
 impl EthereumOpts {
     #[allow(unused)]
-    pub fn provider(&self) -> eyre::Result<Arc<Provider<Http>>> {
-        Ok(Arc::new(Provider::try_from(self.rpc_url.as_str())?))
+    pub async fn signer(&self, chain_id: U256) -> eyre::Result<Option<WalletType>> {
+        self.signer_with(chain_id, Provider::try_from(self.rpc_url.as_str())?).await
     }
 
-    /// Returns a [`LocalWallet`] corresponding to the provided private key or mnemonic
-    pub fn signer(&self) -> eyre::Result<Option<LocalWallet>> {
-        self.wallet.signer()
+    /// Returns a [`SignerMiddleware`] corresponding to the provided private key, mnemonic or hw
+    /// signer
+    pub async fn signer_with(
+        &self,
+        chain_id: U256,
+        provider: Provider<Http>,
+    ) -> eyre::Result<Option<WalletType>> {
+        if self.wallet.ledger {
+            let derivation = HDPath::LedgerLive(self.wallet.mnemonic_index as usize);
```

Comment:
Should this also allow the user to override the full mnemonic path, not just the index? Immediate use case being the Classic format (Vs just having Ledger live)